python/dgl/node_flow.py

```diff
@@ -0,0 +1,408 @@
+                                       readonly=graph_idx.is_readonly())
+        self._parent = parent
+        self._index = graph_idx
+        self._node_mapping = graph_idx.node_mapping
+        self._edge_mapping = graph_idx.edge_mapping
+        self._layer_offsets = graph_idx.layers.tonumpy()
+        self._flow_offsets = graph_idx.flows.tonumpy()
+        self._node_frames = [FrameRef(Frame(num_rows=self.layer_size(i))) for i in range(self.num_layers)]
+        self._edge_frames = [FrameRef(Frame(num_rows=self.flow_size(i))) for i in range(self.num_flows)]
+
+    # override APIs
+    def add_nodes(self, num, data=None):
+        """Add nodes. Disabled because BatchedDGLGraph is read-only."""
+        raise DGLError('Readonly graph. Mutation is not allowed.')
+
+    def add_edge(self, u, v, data=None):
+        """Add one edge. Disabled because BatchedDGLGraph is read-only."""
+        raise DGLError('Readonly graph. Mutation is not allowed.')
+
+    def add_edges(self, u, v, data=None):
+        """Add many edges. Disabled because BatchedDGLGraph is read-only."""
+        raise DGLError('Readonly graph. Mutation is not allowed.')
+
+    def _get_layer_id(self, layer_id):
+        if layer_id >= 0:
+            return layer_id
+        else:
+            return self.num_layers + layer_id
+
+    def _get_flow_id(self, flow_id):
+        if flow_id >= 0:
+            return flow_id
+        else:
+            return self.num_flows + flow_id
+
+    def _get_node_frame(self, layer_id):
+        return self._node_frames[layer_id]
+
+    def _get_edge_frame(self, flow_id):
+        return self._edge_frames[flow_id]
+
+    @property
+    def num_layers(self):
+        """Get the number of layers.
+
+        Returns
+        -------
+        int
+            the number of layers
+        """
+        return len(self._layer_offsets) - 1
+
+    @property
+    def num_flows(self):
+        return self.num_layers - 1
+
+    @property
+    def layers(self):
+        return LayerView(self)
+
+    @property
+    def flows(self):
+        return FlowView(self)
+
+    def layer_size(self, layer_id):
+        """Return the number of nodes in a specified layer."""
+        layer_id = self._get_layer_id(layer_id)
+        return self._layer_offsets[layer_id + 1] - self._layer_offsets[layer_id]
+
+    def flow_size(self, flow_id):
+        flow_id = self._get_flow_id(flow_id)
+        return self._flow_offsets[flow_id + 1] - self._flow_offsets[flow_id]
+
+    def copy_from_parent(self):
+        """Copy node/edge features from the parent graph.
+
+        All old features will be removed.
+        """
+        # TODO we need to avoid copying the same node embedding to all layers.
+        if self._parent._node_frame.num_rows != 0:
+            for i in range(self.num_layers):
+                nid = utils.toindex(self.layer_parent_nid(i))
+                self._node_frames[i] = FrameRef(Frame(self._parent._node_frame[nid]))
```

Comment:
Actually I think there is no additional memory allocation in the current implementation (FrameRef of a Frame creating from an existing frame, since `select_rows` is lazily). When user copy the NodeFrame to gpu externally (like https://github.com/dmlc/dgl/blob/master/examples/mxnet/sse/sse_batch.py#L155), then some keys are wasted.


Could we implement it as

```
def copy_from_parent(self, layer_id_to_keys):
    ...
    self._node_frames[i] = FrameRef(Frame(self._parent._node_frame[nid], layer_id_to_keys[i]))
```

Also slightly modify the the creation of a new frame from an existing frame (adding `keys=None`)

```
class Frame(MutableMapping):
    def __init__(self, data=None, num_rows=0, keys=None):
        ...
        self._columns = {k : Column.create(v) for k, v in data.items() if k in keys} # line 203 in frame.py
```